### DISCUSSION TITLE
How to pass custom CLI arguments to my globalSetup script?

### DISCUSSION THREAD

**Discussion by maxpatiiuk** [vitest-dev/vitest#6593] [Q&A] [answered]
Inside of my global setup file, I am starting Puppeteer browser to do E2E testing (I will migrate to Vitest Browser Mode in the future, but not ready yet).

I would like to add a CLI option to Vitest like `--devtools` that would be used by my global setup script to decide if Puppeteer should be started with DevTools opened or not.

However, if I do so, Vitest will complain about unknown CLI options: "Unknown option `--test`". Are there other ways to pass arguments to globalSetup script?

Options considered:
- Creating a small CLI wrapper around vitest that would convert my CLI argument into an env variable before passing the rest to Vitest's API. This would be a bit of maintenance work, and may make IDE integration harder.
  - Stencil done something similar when they wished to add custom arguments to Jest. From what I have seen this is a big maintenance effort and source of issues (i.e https://github.com/ionic-team/stencil/issues/5640 and https://github.com/ionic-team/stencil/issues/3156).
- Enabling devtools based on whether `--inspect` argument was passed to Vitest. This works, but debugging Node.js test and debugging the rendered browser output via browser DevTools are separate things and sometimes I need to debug just one side or another, rather than both.
- Using env variables (`PUPPETEER_DEBUG=true yarn vitest`). This works, but is more verbose than `yarn vitest --devtools`. It also doesn't work on Windows without crossenv.
  - This might have to be my solution for now, along with some package.json scripts to shorten the command a bit.

Are there better options available that I am not aware of? Thanks!

**Comment by hi-ogawa** [comment, marked as answer]
There were similar requests (the one I remember is on Vite https://github.com/vitejs/vite/issues/16251), but personally it feels unusual to support such custom argument for tools like Vite and Vitest, so probably it's not going forward.

I would imagine your 3rd option (env var) isn't too bad. For pnpm (and probably yarn?), shell emulator provides a uniform env var support https://pnpm.io/cli/run#shell-emulator

**Reply by maxpatiiuk** [reply]
Thanks - I looked for duplicate issues in Vitest, but forgot to look in Vite.
I didn't realize that yarn does shell emulation - neat.
I will proceed with env variables and will mark this discussion as resolved.

**Comment by gladykov** [comment]
+1

Our use case: `npm run generateOutput -- --macro=diagram --only=pdf --single=asciidoc_2_in.txt`

and in package.json `"generateOutput": "vitest utils/outputGenerator.test.ts"`

We setup a lot of things when running tests with `globalSetup` and `customEnvironment`. To generate expected fixture we used same setup, to skip maintaining 2 pipelines.

Our workaround - we write file with args

Command: ` npm run generateOutput -- --macro=diagram --only=pdf --single=asciidoc_2_in.txt`

`package.json`
`"generateOutput": "chmod +x utils/outputGenerator.sh && sh utils/outputGenerator.sh",`


`utils/outputGenerator.sh`

```
while getopts :sc:-: o; do
    case $o in
    :) echo >&2 "option -$OPTARG needs an argument"; continue;;
    '?') echo >&2 "unknown option -$OPTARG"; continue;;
    -) o=${OPTARG%%=*}; OPTARG=${OPTARG#"$o"}; OPTARG=${OPTARG#=};;
    esac
    eval "$o=$OPTARG"
done
shift "$((OPTIND - 1))"

tsx utils/prepareArgsForOutputGenerator.ts --macro=$macro --only=$only --single=$single --theme=$theme // this will save values to file
vitest utils/outputGenerator.test.ts  // this will read values from file
```
